enhancements/ingress/add-headers-to-haproxyconfig-via-ingress-controller-or-route-api.md

```diff
@@ -0,0 +1,295 @@
+environment variables, `INGRESS_SET_HTTP_HEADERS`, which specify the headers to set in the 
+global section of the haproxy.config.
+The ingress operator sets this variable based on the IngressController's
+`spec.httpSetHeaders` field value.
+
+The values of `INGRESS_SET_HTTP_HEADERS` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
+
+For example, if `ROUTER_SET_HTTP_HEADERS` is set to
+`X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block`, then the template adds the following stanzas to
+the appropriate frontends:
+
+```haproxy.config's global section
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+This feature's implementation will be similar to an existing feature called logging headers which was implemented in 
+[router](https://github.com/openshift/router/commit/40ee0ba472a4ac503f1e6c97d4464f559101c97c) and 
+[cluster-ingress-operator](https://github.com/openshift/cluster-ingress-operator/commit/c5139cb07c3a99a8a3b18a6c9321310add51b075).
+
+B) Setting HTTP headers in the backend section `default_backend openshift_default` of the haproxy.config via Route's Spec.
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/router
+
+The router configures HAProxy using a configuration template. The template is populated with values in ServiceAliasConfig .
+The enhancement adds a field in ServiceAliasConfig called `RouteSetHTTPHeaders`, which specify the headers to set in the
+backend section `default_backend openshift_default` section of the haproxy.config.
+The router sets this variable based on the Route's
+`spec.httpSetHeaders` field value. Service AliasConfig will be populated at this [code location](https://github.com/openshift/router/blob/master/pkg/router/template/router.go#L927)
+
+The values of `RouteSetHTTPHeaders` are comma-delimited lists of values of
+the form `name:value`.  (Colons and commas are not allowed in HTTP header
+names, so their use unambiguously delimits values in the environment variable.)
+The template translates these values into an appropriate `http-response set-header` stanzas.
+
+For example, if `RouteSetHTTPHeaders` is set to
+`X-Frame-Options: DENY,X-XSS-Protection: 1;mode=block`, then the template adds the following stanzas to
+the appropriate frontends:
+
+```haproxy.config's backend section called default_backend openshift_default
+    http-response set-header X-Frame-Options: DENY
+    http-response set-header X-XSS-Protection: 1;mode=block    
+```
+
+[Code location](https://github.com/openshift/router/blob/master/images/router/haproxy/conf/haproxy-config.template#L608) where this
+will be added.
+
+### Risks and Mitigations
```

Comment:
I'm wondering how this feature will co-exist with [the OpenShift's Ingress HSTS policy](https://github.com/openshift/enhancements/blob/master/enhancements/ingress/global-admission-hsts.md) which also sets a header (STS header) on the routes.

Reply:
This is addressed by a comment in the RFE:

> Create an “immutable” OpenShift deny-list of headers that are known in advance to be problematic (e.g. HSTS)

https://issues.redhat.com/browse/RFE-464?focusedCommentId=20064927&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-20064927

Reply:
@Miciah but won't it be a different epic ?

Reply:
In the router we can check the header for `Strict-Transport-Security:` in the route and then ignore it.

Reply:
Why would it be a different epic?  

Going back to the comment I cited:

> Create an “immutable” OpenShift deny-list of headers that are known in advance to be problematic (e.g. HSTS)

We don't need an API field for the immutable deny-list, which would include `strict-transport-security`.

The comment I cited also mentions the following:

> Cluster Admins can delegate per-route controls (perhaps with an allow-list, or an attribute in the global ones above) to, for example, Project Admins

We would need a field in the IngressController API for that.  If there's a reason this should be a separate epic, please give an argument.  We can discuss it with PM if it really should be out of scope.

Reply:
I see, so the immutable deny list will exist in the code and will be documented. It'll include HSTS (as managed by the route's annotation) and potentially some other headers.

Reply:
@Miciah : if the deny list is part of the EP, what about the other statement from the same comment:

> We will allow global replace, delete and append options.

The way @miheer is suggestion is via `httpSetHeaders` field which looks like a mix of replace and append. However the delete is something I'm wondering about: what the delete header list is supposed to mean on the ingresscontroller and route levels? How the conflict between the deleted and set headers should be managed? Should the pre-defined (existed before, added by some route config option) headers be impacted or we have to add them all to the deny list?

Reply:
list added